### DISCUSSION TITLE
Data augmentation causes 'DataLoader worker (pid(s) 2642711) exited unexpectedly'

### DISCUSSION THREAD

**Discussion by dianemarquette** [Project-MONAI/MONAI#2404] [Q&A]
Hi !

I used MONAI's 3D spleen segmentation tutorial to train a model with my own dataset. However, I would like to leverage data augmentation to reduce overfitting. 

When I define my preprocessing transforms for my training dataset like in the tutorial:
```
train_transforms = Compose(
                [
                    # load the NIfTI files
                    LoadImaged(keys=["image", "label"]),
                    # convert image to "channel-first" format
                    AddChanneld(keys=["image", "label"]),
                    # resample to a consistent voxel format
                    Spacingd(
                        keys=["image", "label"],
                        pixdim=self.pix_dim, 
                        mode=("bilinear", "nearest"),
                    ),
                    # reorientate volumes to have a consistent axes orientation
                    Orientationd(keys=["image", "label"], axcodes="RAS"),
                    
                    ScaleIntensityRanged(
                        keys=["image"], a_min=0, a_max=2000,
                        b_min=0.0, b_max=1.0, clip=True,
                    ),
                    
                    CropForegroundd(keys=["image", "label"], source_key="image"),
                    # randomly crop out patch samples from
                    # big image based on pos / neg ratio
                    # the image centers of negative samples
                    # must be in valid image area
                    RandCropByPosNegLabeld(
                        keys=["image", "label"],
                        label_key="label",
                        spatial_size=self.spatial_size,
                        pos=1,
                        neg=1,
                        num_samples=4,
                        image_key="image",
                        image_threshold=0,
                    ),
                    ToTensord(keys=["image", "label"]),
                ]
            )
```
everything runs smoothly. 

However, when I include additional transforms for data augmentation:
```
train_transforms = Compose(
                [
                    # load the NIfTI files
                    LoadImaged(keys=["image", "label"]),
                    # convert image to "channel-first" format
                    AddChanneld(keys=["image", "label"]),
                    # resample to a consistent voxel format
                    Spacingd(
                        keys=["image", "label"],
                        pixdim=self.pix_dim, 
                        mode=("bilinear", "nearest"),
                    ),
                    # reorientate volumes to have a consistent axes orientation
                    Orientationd(keys=["image", "label"], axcodes="RAS"),
                    
                    ScaleIntensityRanged(
                        keys=["image"], a_min=0, a_max=2000,
                        b_min=0.0, b_max=1.0, clip=True,
                    ),
                    
                    # flip axis to train model for left and right hands
                    RandFlipd(keys=["image", "label"], prob=0.1, spatial_axis=0),
                    RandFlipd(keys=["image", "label"], prob=0.1, spatial_axis=1),
                    
                    RandAffined(keys=["image", "label"], 
                        mode=("bilinear", "nearest"), prob=0.1,
                        rotate_range=(np.pi / 36, np.pi / 36, np.pi * 2), 
                        translate_range=(40, 40, 2),
                        padding_mode="zeros",
                    ),
                    RandAffined(keys=["image", "label"], 
                        mode=("bilinear", "nearest"), prob=0.1,
                        rotate_range=(np.pi / 36, np.pi / 2, np.pi / 36), 
                        translate_range=(40, 40, 2),
                        padding_mode="zeros",
                    ),
                    RandAffined(keys=["image", "label"],
                        mode=("bilinear", "nearest"), prob=0.1,
                        rotate_range=(np.pi / 2, np.pi / 36, np.pi / 36),
                        translate_range=(40, 40, 2), 
                        padding_mode="zeros",
                    ),
                    
                    # simulate trapezium deformations 
                    Rand3DElasticd(keys=["image", "label"], 
                        mode=("bilinear", "nearest"), prob=0.1,
                        sigma_range=(5, 8), magnitude_range=(100, 200), 
                        scale_range=(0.15, 0.15, 0.15), padding_mode="zeros",
                    ),

                    RandZoomd(keys=["image", "label"],  
                        max_zoom=1.1, min_zoom=0.9, 
                        mode=('trilinear', 'nearest'),
                        align_corners=(True, None),
                        prob=0.1
                    ),

                    CropForegroundd(keys=["image", "label"], source_key="image"),
                    
                    # randomly crop out patch samples from
                    # big image based on pos / neg ratio
                    # the image centers of negative samples
                    # must be in valid image area
                    RandCropByPosNegLabeld(
                        keys=["image", "label"],
                        label_key="label",
                        spatial_size=self.spatial_size,
                        pos=1,
                        neg=1,
                        num_samples=4,
                        image_key="image",
                        image_threshold=0,
                    ),
                    RandGaussianNoised(keys=["image"], prob=0.1, 
                        mean=0.0, std=0.1,
                    ),
                    RandShiftIntensityd(keys=["image"], 
                        offsets=np.random.uniform(0,0.3), prob=0.1,
                    ),
                    ToTensord(keys=["image", "label"]),
                ]
            )
```
I get the following error:
```
Traceback (most recent call last):
  File "/home/diane/miniconda3/envs/lightning-hydra/lib/python3.8/site-packages/torch/utils/data/dataloader.py", line 986, in _try_get_data
    data = self._data_queue.get(timeout=timeout)
  File "/home/diane/miniconda3/envs/lightning-hydra/lib/python3.8/multiprocessing/queues.py", line 107, in get
    if not self._poll(timeout):
  File "/home/diane/miniconda3/envs/lightning-hydra/lib/python3.8/multiprocessing/connection.py", line 257, in poll
    return self._poll(timeout)
  File "/home/diane/miniconda3/envs/lightning-hydra/lib/python3.8/multiprocessing/connection.py", line 424, in _poll
    r = wait([self], timeout)
  File "/home/diane/miniconda3/envs/lightning-hydra/lib/python3.8/multiprocessing/connection.py", line 931, in wait
    ready = selector.select(timeout)
  File "/home/diane/miniconda3/envs/lightning-hydra/lib/python3.8/selectors.py", line 415, in select
    fd_event_list = self._selector.poll(timeout)
  File "/home/diane/miniconda3/envs/lightning-hydra/lib/python3.8/site-packages/torch/utils/data/_utils/signal_handling.py", line 66, in handler
    _error_if_any_worker_fails()
RuntimeError: DataLoader worker (pid 2643489) is killed by signal: Killed. 

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "run_keri.py", line 35, in main
    return train(config)
  File "/media/diane/Shared Data/KeriNet/src/train.py", line 77, in train
    trainer.fit(model=model, datamodule=datamodule)
  File "/home/diane/miniconda3/envs/lightning-hydra/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 458, in fit
    self._run(model)
  File "/home/diane/miniconda3/envs/lightning-hydra/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 756, in _run
    self.dispatch()
  File "/home/diane/miniconda3/envs/lightning-hydra/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 797, in dispatch
    self.accelerator.start_training(self)
  File "/home/diane/miniconda3/envs/lightning-hydra/lib/python3.8/site-packages/pytorch_lightning/accelerators/accelerator.py", line 96, in start_training
    self.training_type_plugin.start_training(trainer)
  File "/home/diane/miniconda3/envs/lightning-hydra/lib/python3.8/site-packages/pytorch_lightning/plugins/training_type/training_type_plugin.py", line 144, in start_training
    self._results = trainer.run_stage()
  File "/home/diane/miniconda3/envs/lightning-hydra/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 807, in run_stage
    return self.run_train()
  File "/home/diane/miniconda3/envs/lightning-hydra/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 869, in run_train
    self.train_loop.run_training_epoch()
  File "/home/diane/miniconda3/envs/lightning-hydra/lib/python3.8/site-packages/pytorch_lightning/trainer/training_loop.py", line 481, in run_training_epoch
    for batch_idx, (batch, is_last_batch) in train_dataloader:
  File "/home/diane/miniconda3/envs/lightning-hydra/lib/python3.8/site-packages/pytorch_lightning/profiler/profilers.py", line 112, in profile_iterable
    value = next(iterator)
  File "/home/diane/miniconda3/envs/lightning-hydra/lib/python3.8/site-packages/pytorch_lightning/trainer/supporters.py", line 534, in prefetch_iterator
    for val in it:
  File "/home/diane/miniconda3/envs/lightning-hydra/lib/python3.8/site-packages/pytorch_lightning/trainer/supporters.py", line 464, in __next__
    return self.request_next_batch(self.loader_iters)
  File "/home/diane/miniconda3/envs/lightning-hydra/lib/python3.8/site-packages/pytorch_lightning/trainer/supporters.py", line 478, in request_next_batch
    return apply_to_collection(loader_iters, Iterator, next)
  File "/home/diane/miniconda3/envs/lightning-hydra/lib/python3.8/site-packages/pytorch_lightning/utilities/apply_func.py", line 84, in apply_to_collection
    return function(data, *args, **kwargs)
  File "/home/diane/miniconda3/envs/lightning-hydra/lib/python3.8/site-packages/torch/utils/data/dataloader.py", line 517, in __next__
    data = self._next_data()
  File "/home/diane/miniconda3/envs/lightning-hydra/lib/python3.8/site-packages/torch/utils/data/dataloader.py", line 1182, in _next_data
    idx, data = self._get_data()
  File "/home/diane/miniconda3/envs/lightning-hydra/lib/python3.8/site-packages/torch/utils/data/dataloader.py", line 1148, in _get_data
    success, data = self._try_get_data()
  File "/home/diane/miniconda3/envs/lightning-hydra/lib/python3.8/site-packages/torch/utils/data/dataloader.py", line 999, in _try_get_data
    raise RuntimeError('DataLoader worker (pid(s) {}) exited unexpectedly'.format(pids_str)) from e
RuntimeError: DataLoader worker (pid(s) 2643489) exited unexpectedly
```
I've tried to fix it, but I'm a bit stuck. Any guidance / tip would be greatly appreciated :).

**Comment by Nic-Ma** [comment]
Hi @dianemarquette ,

Thanks for your interest here.
I think maybe your transforms are too heavy for DataLoaders, could you please help try smaller `num_workers` or `num_workers=0` in the DataLoader?
https://github.com/pytorch/pytorch/blob/master/torch/utils/data/dataloader.py#L691

Thanks.

**Reply by dianemarquette** [reply]
@Nic-Ma Thanks for the quick response. 

I tried using a smaller `num_workers` such as 1 and 2. Unfortunately, I got the exact same error. 

I also tried setting `num_workers` to 0. With this setup, my python script gets killed! 
```
Loading dataset: 100%|██████████████████████████| 23/23 [01:36<00:00,  4.20s/it]
Epoch 0:   0%|                                           | 0/21 [00:00<?, ?it/s]
[1]    2644250 killed     python run_keri.py datamodule.data_augmentation=True datamodule.num_workers=0
```
You mentioned that my transforms could be too heavy for DataLoaders. What is the maximum number of MONAI transforms that you recommend to compose?

**Reply by Nic-Ma** [reply]
Actually, we don't have a recommended transform number, maybe you can try to use regular `Dataset` instead of `CacheDataset` and smaller num_workers for DataLoader.

Thanks.

**Reply by rijobro** [reply]
Hi @dianemarquette, I'm also curious as to why it worked with your initial transforms, and then stopped working as you added in more transforms. One thing to note is that I think some of your transforms are redundant. Multiple `RandAffined` transforms could be simply 1 transform (with a larger range), and this would decrease image degradation through from multiple image resamples. Further, the `Rand3DElasticd` is also capable of performing affine transformations at the same time, so I think you could remove all uses `RandAffined`. I also think the `RandZoomd` could also be incorporated into `Rand3DElasticd` with `scale_range`.

This wouldn't be the same as your current set up, since you have separate probabilities for each of your independent transforms, but it might be a good starting place.

**Reply by dianemarquette** [reply]
@Nic-Ma Using regular `Dataset` instead of `CacheDataset` with `num_workers` = 2 stopped the training from crashing at the beginning of the first epoch. However, the training is so much slower ! It now takes almost 4 minutes to complete a single epoch. Plus, why did `Dataset` instead of `CacheDataset` make such a difference?

@rijobro Thanks for your input! This redundancy was actually desired. I wanted to generate images slightly transformed with the probability of having a single or a random combination of independent transforms applied to it. However, because this large number of transforms seems to be problematic, I will try out your suggested implementation.

**Reply by Nic-Ma** [reply]
Hi @dianemarquette ,

`CacheDataset` uses more memory to store the cache content, which may also cause program to be killed.

Thanks.

**Reply by dianemarquette** [reply]
@Nic-Ma Ok, I see. Is there any other way than `CacheDataset` to speed up the training when using `Dataset`?

**Reply by Nic-Ma** [reply]
You can consider to use `PersistentDataset` which stores data in your filesystem instead of memory.

Thanks.

**Reply by dianemarquette** [reply]
@rijobro I've experimented a bit more with different transform compositions. Combining the standard preprocessing transforms from the spleen 3D segmentation [tutorial](https://github.com/Project-MONAI/tutorials/blob/master/3d_segmentation/spleen_segmentation_3d.ipynb) with `RandFlipd`, `RandZoomd`, `RandGaussianNoised`, and `RandShiftIntensityd` works great. However, as soon as I add `Rand3DElasticd` or even if it's the only transform that I use for data augmentation, the DataLoader is overwhelmed or I get an OOM error or the script gets killed. Is `Rand3DElasticd` significantly memory intensive? Must it be only used with specific transforms?

I read in Discussion #2440 that "the multi-worker loader doesn't work well with the GPU-based transforms". Is it only due to additional CPU-GPU communication overheads or are there other factors at play?

**Reply by rijobro** [reply]
Just in case it's a GPU/multi-process problem – I think that `Rand3dElasticd` uses the CPU by default (`device=None`), are you explicitly setting e.g., `device=cuda`? If you are, try removing it. If you're not then I don't think it's a GPU/multi-process problem.

However, I think that it's more likely to be a problem with your input parameters. If you look at the [test uses](https://github.com/Project-MONAI/MONAI/blob/dev/tests/test_rand_elasticd_3d.py) of `Rand3dElastic`, the magnitudes are much smaller than the ones you've used (~0.3 instead of 100). I haven't looked at the implementation, but if large magnitudes result in large deformation grids, this could explain why you're running out of memory. I would try with smaller arguments, visualise the results, and gradually increase until you're happy. You don't need to do these tests using datasets or data loaders (avoids the confusion of multiprocessing etc). In a jupyter notebook, you should be able to do something like:

```python
image_fname = ...
transforms = Compose([
    LoadImage(image_only=True),
    Rand3DElastic(test_parameters),
])
im = transforms(image_fname)
# do some matplotlib visualising
```

**Reply by dianemarquette** [reply]
@rijobro Thanks for the catch regarding the magnitude that was indeed too high. I reduced the magnitude range to 0.3-2.3 as used in MONAI's tests. However, it still didn't solve my problem. It works in a Jupyter notebook when I only load a small training dataset (e.g., 2 CT scans). However, if I load the full dataset (23 scans), the kernel crashes. I don't understand why it occurs if it doesn't use the GPU. Any ideas?

**Reply by rijobro** [reply]
Does the error occur when you use `Dataset` instead of `CacheDataset`? If not, it may be easiest to just use that. 

As a half-way house, you could also use `PersistentDataset`, which will pre-compute your deterministic transforms (LoadImage, AddChannel, Spacing, Orientation, ScaleIntensity), save them to file and then only compute the random transforms each time an element in the dataset is accessed.

**Reply by dianemarquette** [reply]
@rijobro The error does not occur when I use `Dataset` in my Jupyter notebook dummy example, but the dataloader exits unexpectedly when I run it in the script training my model. 

I've tried using  `PersistentDataset` with `num_workers=1`. It works in my Jupyter notebook, but I get the following error when I include it in my script:
```
RuntimeError: [enforce fail at CPUAllocator.cpp:67] . DefaultCPUAllocator: can't allocate memory: you tried to allocate 37613947968 bytes. Error code 12 (Cannot allocate memory)

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/diane/miniconda3/envs/lightning-hydra/lib/python3.8/site-packages/torch/utils/data/_utils/worker.py", line 202, in _worker_loop
    data = fetcher.fetch(index)
  File "/home/diane/miniconda3/envs/lightning-hydra/lib/python3.8/site-packages/torch/utils/data/_utils/fetch.py", line 44, in fetch
    data = [self.dataset[idx] for idx in possibly_batched_index]
  File "/home/diane/miniconda3/envs/lightning-hydra/lib/python3.8/site-packages/torch/utils/data/_utils/fetch.py", line 44, in <listcomp>
    data = [self.dataset[idx] for idx in possibly_batched_index]
  File "/home/diane/miniconda3/envs/lightning-hydra/lib/python3.8/site-packages/monai/data/dataset.py", line 94, in __getitem__
    return self._transform(index)
  File "/home/diane/miniconda3/envs/lightning-hydra/lib/python3.8/site-packages/monai/data/dataset.py", line 267, in _transform
    return self._post_transform(pre_random_item)
  File "/home/diane/miniconda3/envs/lightning-hydra/lib/python3.8/site-packages/monai/data/dataset.py", line 217, in _post_transform
    item_transformed = apply_transform(_transform, item_transformed)
  File "/home/diane/miniconda3/envs/lightning-hydra/lib/python3.8/site-packages/monai/transforms/transform.py", line 73, in apply_transform
    raise RuntimeError(f"applying transform {transform}") from e
RuntimeError: applying transform <monai.transforms.spatial.dictionary.Rand3DElasticd object at 0x7fb209ff6310>
```
I also tried using `PersistentDataset` with `num_workers=0`. The script gets killed:
`Epoch 0:   0%|                                           | 0/21 [00:00<?, ?it/s][1]    2667243 killed     python run_keri.py datamodule.num_workers=0
`

I don't know what to do. Maybe I just don't have enough RAM in my GPU to apply this transform to my training set. It's quite frustrating.

**Reply by rijobro** [reply]
Ok, from your message, I get the impression that you've tried two of the following four possibilities, is that correct? If so, could you try the other two? (I'm trying to figure out if there's something going wrong in the DataLoader step.)

- jupyter using dataset: WORKS
- jupyter using dataset and dataloader: ??
- script using dataset: ??
- script using dataset and dataloader: DOESN'T WORK

Are the images the same size in both examples? We have functions `create_test_image_2d` and `create_test_image_3d` to create random images if this would help you create these tests. 

If you create a minimum working example (one that doesn't use your data but does work in jupyter), then feel free to send this and I'll have a look to see if there's anything obvious.

As you say, if we can't find any bugs, it could be because the specifications of your computer aren't up to the task, but this seems unlikely to me (a 3d deformation of an image isn't that intensive).

**Comment by jjhhan** [comment]
i have the same error

**Reply by rijobro** [reply]
please provide an example that demonstrates your problem. Thanks

**Comment by jjhhan** [comment]
&nbsp;The&nbsp; content is a routine&nbsp; dataloader and training. The Compose of transform is all i used.&nbsp;
&nbsp; My computer has 64GB ram and a RTX3090
&nbsp;I update to Monai0.9.1 and delete most of transforms. Only keep&nbsp;LoadImaged EnsureChannelFirstd RandCropByLabelClassesd\Orientationd\NormalizeIntensityd.
&nbsp;It works but the dice of val is very low. Only 0.7-0.8&nbsp;


------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "904604254"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2022年9月8日(星期四) 上午9:32
***@***.******@***.***&gt;;
***@***.***&gt;;
主题:&nbsp;回复： [Project-MONAI/MONAI] Data augmentation causes 'DataLoader worker (pid(s) 2642711) exited unexpectedly' (Discussion #2404)



 train_images = sorted(
    glob.glob(os.path.join(data_dir, "imagesTr", "*.nii.gz")))
train_labels = sorted(
    glob.glob(os.path.join(data_dir, "labelsTr", "*.nii.gz")))
data_dicts = [
    {"image": image_name, "label": label_name}
    for image_name, label_name in zip(train_images, train_labels)
]
train_dicts, val_dicts = data_dicts[:-7], data_dicts[-10:]

set_determinism(seed=666)
print('Number of training images per epoch:', len(train_dicts))
print('Number of validating images per epoch:', len(val_dicts))


# Creation of data directories for data_loader
# Transforms for training and validation


train_transforms = [
        LoadImaged(keys=['image', 'label']),
        AddChanneld(keys=['image', 'label']),
        # CropForegroundd(keys=['image', 'label'], source_key='image'),               # crop CropForeground
        # Spacingd(
        #     keys=["image", "label"],
        #     pixdim=(1.5, 1.5, 2.0),
        #     mode=("bilinear", "nearest"),
        # ),
        RandFlipd(keys=['image', 'label'], prob=0.15, spatial_axis=1),
        RandFlipd(keys=['image', 'label'], prob=0.15, spatial_axis=0),
        RandFlipd(keys=['image', 'label'], prob=0.15, spatial_axis=2),
        RandAffined(keys=['image', 'label'], mode=('bilinear', 'nearest'), prob=0.1,
                    rotate_range=(np.pi / 36, np.pi / 36, np.pi * 2), padding_mode="zeros"),
        RandAffined(keys=['image', 'label'], mode=('bilinear', 'nearest'), prob=0.1,
                    rotate_range=(np.pi / 36, np.pi / 2, np.pi / 36), padding_mode="zeros"),
        RandAffined(keys=['image', 'label'], mode=('bilinear', 'nearest'), prob=0.1,
                    rotate_range=(np.pi / 2, np.pi / 36, np.pi / 36), padding_mode="zeros"),
        Rand3DElasticd(keys=['image', 'label'], mode=('bilinear', 'nearest'), prob=0.1,
                       sigma_range=(5, 8), magnitude_range=(100, 200), scale_range=(0.15, 0.15, 0.15),
                       padding_mode="zeros"),
        RandGaussianSmoothd(keys=["image"], sigma_x=(0.5, 1.15), sigma_y=(0.5, 1.15), sigma_z=(0.5, 1.15), prob=0.1,),
        RandAdjustContrastd(keys=['image'], gamma=(0.5, 2.5), prob=0.1),
        RandGaussianNoised(keys=['image'], prob=0.1, mean=np.random.uniform(0, 0.5), std=np.random.uniform(0, 1)),

        SpatialPadd(keys=['image', 'label'], spatial_size=opt.patch_size, method= 'end'),  # pad if the image is smaller than patch
        RandSpatialCropd(keys=['image', 'label'], roi_size=opt.patch_size, random_size=False),
        ToTensord(keys=['image', 'label'])
    ]

val_transforms = [
        LoadImaged(keys=['image', 'label']),
        AddChanneld(keys=['image', 'label']),
        # CropForegroundd(keys=['image', 'label'], source_key='image'),                   # crop CropForeground
        # Spacingd(
        #     keys=["image", "label"],
        #     pixdim=(1.5, 1.5, 2.0),
        #     mode=("bilinear", "nearest"),
        # ),
        SpatialPadd(keys=['image', 'label'], spatial_size=opt.patch_size, method= 'end'),  # pad if the image is smaller than patch
        ToTensord(keys=['image', 'label'])
    ]

train_transforms = Compose(train_transforms)
val_transforms = Compose(val_transforms)


train_ds = monai.data.Dataset(data=train_dicts, transform=train_transforms)
train_loader = DataLoader(train_ds, batch_size=opt.batch_size, shuffle=True, collate_fn=list_data_collate, num_workers=opt.workers, pin_memory=True)


val_ds = monai.data.Dataset(data=val_dicts, transform=val_transforms)
val_loader = DataLoader(val_ds, batch_size=1, num_workers=opt.workers, collate_fn=list_data_collate, pin_memory=True)
post_pred = Compose([EnsureType(), AsDiscrete(argmax=True, to_onehot=6), KeepLargestConnectedComponent(is_onehot=True, applied_labels=[2,4,5])])
post_label = Compose([EnsureType(), AsDiscrete(to_onehot=6)])
saver_ori = SaveImage(output_dir=opt.output_folder, output_ext=".nii.gz", output_postfix="ori",print_log=True)
saver_gt = SaveImage(output_dir=opt.output_folder, output_ext=".nii.gz", output_postfix="gt",print_log=True)
saver_seg = SaveImage(output_dir=opt.output_folder, output_ext=".nii.gz", output_postfix="seg",print_log=True)



net = build_UNETR() # UneTR
print('net == unetr')
net.to(device)

# loss
loss_function = monai.losses.DiceCELoss(to_onehot_y=True, softmax=True)
torch.backends.cudnn.benchmark = opt.benchmark  # for accelerating Convs
dice_metric = DiceMetric(include_background=True, reduction="mean", get_not_nans=False)  # Mean Dice caculate
dice_metric_batch = DiceMetric(include_background=True, reduction="mean_batch")
scaler = torch.cuda.amp.GradScaler()


optim = torch.optim.AdamW(net.parameters(), lr=1e-4, weight_decay=1e-5)



val_interval = 1
best_metric = -1
best_metric_epoch = -1
epoch_loss_values = []
metric_values = []
metric_values_tb = []
metric_values_tc = []
metric_values_fb = []
metric_values_lfc = []
metric_values_rfc = []
train_time = time()
writer = SummaryWriter() # store in runs/***


for epoch in range(opt.epochs):
    print("-" * 20)
    epoch_start = time()
    net.train()
    epoch_loss = 0
    step = 0
    pbar = tqdm(train_loader, dynamic_ncols=True)
    for batch_data in pbar:
        pbar.set_description(desc='Train_Epoch : {}/{}'.format(epoch+1, opt.epochs))
        step += 1
        inputs, labels = (
            batch_data["image"].to(device),
            batch_data["label"].to(device),
        )
        with torch.cuda.amp.autocast():
            logit_map = net(inputs)
            loss = loss_function(logit_map, labels)
        # outputs = net(inputs)
        # loss = loss_function(outputs, labels)
        scaler.scale(loss).backward()
        epoch_loss += loss.item()
        scaler.unscale_(optim)
        scaler.step(optim)
        scaler.update()
        optim.zero_grad()


    epoch_loss /= step
    epoch_loss_values.append(epoch_loss)
    print(f"epoch {epoch + 1} average loss: {epoch_loss:.4f}  ")


&nbsp;


904604254
***@***.***



&nbsp;




------------------ 原始邮件 ------------------
发件人: "Richard ***@***.***&gt;; 
发送时间: 2022年9月1日(星期四) 凌晨0:15
收件人: ***@***.***&gt;; 
抄送: ***@***.***&gt;; ***@***.***&gt;; 
主题: Re: [Project-MONAI/MONAI] Data augmentation causes 'DataLoader worker (pid(s) 2642711) exited unexpectedly' (Discussion #2404)





 
please provide an example that demonstrates your problem. Thanks

**Comment by jjhhan** [comment]
train_images = sorted(
    glob.glob(os.path.join(data_dir, "imagesTr", "*.nii.gz")))
train_labels = sorted(
    glob.glob(os.path.join(data_dir, "labelsTr", "*.nii.gz")))
data_dicts = [
    {"image": image_name, "label": label_name}
    for image_name, label_name in zip(train_images, train_labels)
]
train_dicts, val_dicts = data_dicts[:-7], data_dicts[-10:]

set_determinism(seed=666)
print('Number of training images per epoch:', len(train_dicts))
print('Number of validating images per epoch:', len(val_dicts))


# Creation of data directories for data_loader
# Transforms for training and validation


train_transforms = [
        LoadImaged(keys=['image', 'label']),
        AddChanneld(keys=['image', 'label']),
        # CropForegroundd(keys=['image', 'label'], source_key='image'),               # crop CropForeground
        # Spacingd(
        #     keys=["image", "label"],
        #     pixdim=(1.5, 1.5, 2.0),
        #     mode=("bilinear", "nearest"),
        # ),
        RandFlipd(keys=['image', 'label'], prob=0.15, spatial_axis=1),
        RandFlipd(keys=['image', 'label'], prob=0.15, spatial_axis=0),
        RandFlipd(keys=['image', 'label'], prob=0.15, spatial_axis=2),
        RandAffined(keys=['image', 'label'], mode=('bilinear', 'nearest'), prob=0.1,
                    rotate_range=(np.pi / 36, np.pi / 36, np.pi * 2), padding_mode="zeros"),
        RandAffined(keys=['image', 'label'], mode=('bilinear', 'nearest'), prob=0.1,
                    rotate_range=(np.pi / 36, np.pi / 2, np.pi / 36), padding_mode="zeros"),
        RandAffined(keys=['image', 'label'], mode=('bilinear', 'nearest'), prob=0.1,
                    rotate_range=(np.pi / 2, np.pi / 36, np.pi / 36), padding_mode="zeros"),
        Rand3DElasticd(keys=['image', 'label'], mode=('bilinear', 'nearest'), prob=0.1,
                       sigma_range=(5, 8), magnitude_range=(100, 200), scale_range=(0.15, 0.15, 0.15),
                       padding_mode="zeros"),
        RandGaussianSmoothd(keys=["image"], sigma_x=(0.5, 1.15), sigma_y=(0.5, 1.15), sigma_z=(0.5, 1.15), prob=0.1,),
        RandAdjustContrastd(keys=['image'], gamma=(0.5, 2.5), prob=0.1),
        RandGaussianNoised(keys=['image'], prob=0.1, mean=np.random.uniform(0, 0.5), std=np.random.uniform(0, 1)),

        SpatialPadd(keys=['image', 'label'], spatial_size=opt.patch_size, method= 'end'),  # pad if the image is smaller than patch
        RandSpatialCropd(keys=['image', 'label'], roi_size=opt.patch_size, random_size=False),
        ToTensord(keys=['image', 'label'])
    ]

val_transforms = [
        LoadImaged(keys=['image', 'label']),
        AddChanneld(keys=['image', 'label']),
        # CropForegroundd(keys=['image', 'label'], source_key='image'),                   # crop CropForeground
        # Spacingd(
        #     keys=["image", "label"],
        #     pixdim=(1.5, 1.5, 2.0),
        #     mode=("bilinear", "nearest"),
        # ),
        SpatialPadd(keys=['image', 'label'], spatial_size=opt.patch_size, method= 'end'),  # pad if the image is smaller than patch
        ToTensord(keys=['image', 'label'])
    ]

train_transforms = Compose(train_transforms)
val_transforms = Compose(val_transforms)


train_ds = monai.data.Dataset(data=train_dicts, transform=train_transforms)
train_loader = DataLoader(train_ds, batch_size=opt.batch_size, shuffle=True, collate_fn=list_data_collate, num_workers=opt.workers, pin_memory=True)


val_ds = monai.data.Dataset(data=val_dicts, transform=val_transforms)
val_loader = DataLoader(val_ds, batch_size=1, num_workers=opt.workers, collate_fn=list_data_collate, pin_memory=True)
post_pred = Compose([EnsureType(), AsDiscrete(argmax=True, to_onehot=6), KeepLargestConnectedComponent(is_onehot=True, applied_labels=[2,4,5])])
post_label = Compose([EnsureType(), AsDiscrete(to_onehot=6)])
saver_ori = SaveImage(output_dir=opt.output_folder, output_ext=".nii.gz", output_postfix="ori",print_log=True)
saver_gt = SaveImage(output_dir=opt.output_folder, output_ext=".nii.gz", output_postfix="gt",print_log=True)
saver_seg = SaveImage(output_dir=opt.output_folder, output_ext=".nii.gz", output_postfix="seg",print_log=True)



net = build_UNETR() # UneTR
print('net == unetr')
net.to(device)

# loss
loss_function = monai.losses.DiceCELoss(to_onehot_y=True, softmax=True)
torch.backends.cudnn.benchmark = opt.benchmark  # for accelerating Convs
dice_metric = DiceMetric(include_background=True, reduction="mean", get_not_nans=False)  # Mean Dice caculate
dice_metric_batch = DiceMetric(include_background=True, reduction="mean_batch")
scaler = torch.cuda.amp.GradScaler()


optim = torch.optim.AdamW(net.parameters(), lr=1e-4, weight_decay=1e-5)



val_interval = 1
best_metric = -1
best_metric_epoch = -1
epoch_loss_values = []
metric_values = []
metric_values_tb = []
metric_values_tc = []
metric_values_fb = []
metric_values_lfc = []
metric_values_rfc = []
train_time = time()
writer = SummaryWriter() # store in runs/***


for epoch in range(opt.epochs):
    print("-" * 20)
    epoch_start = time()
    net.train()
    epoch_loss = 0
    step = 0
    pbar = tqdm(train_loader, dynamic_ncols=True)
    for batch_data in pbar:
        pbar.set_description(desc='Train_Epoch : {}/{}'.format(epoch+1, opt.epochs))
        step += 1
        inputs, labels = (
            batch_data["image"].to(device),
            batch_data["label"].to(device),
        )
        with torch.cuda.amp.autocast():
            logit_map = net(inputs)
            loss = loss_function(logit_map, labels)
        # outputs = net(inputs)
        # loss = loss_function(outputs, labels)
        scaler.scale(loss).backward()
        epoch_loss += loss.item()
        scaler.unscale_(optim)
        scaler.step(optim)
        scaler.update()
        optim.zero_grad()


    epoch_loss /= step
    epoch_loss_values.append(epoch_loss)
    print(f"epoch {epoch + 1} average loss: {epoch_loss:.4f}  ")


&nbsp;


904604254
***@***.***



&nbsp;




------------------&nbsp;原始邮件&nbsp;------------------
发件人: "Richard ***@***.***&gt;; 
发送时间: 2022年9月1日(星期四) 凌晨0:15
收件人: ***@***.***&gt;; 
抄送: ***@***.***&gt;; ***@***.***&gt;; 
主题: Re: [Project-MONAI/MONAI] Data augmentation causes &#39;DataLoader worker (pid(s) 2642711) exited unexpectedly&#39; (Discussion #2404)





 
please provide an example that demonstrates your problem. Thanks
